project-management/src/main/java/life/qbic/projectmanagement/application/api/AsyncProjectServiceImpl.java

```diff
@@ -318,7 +331,126 @@ public Mono<DigitalObject> sampleInformationTemplate(String projectId, String ex
   @Override
   public Flux<ValidationResponse> validate(Flux<ValidationRequest> requests)
       throws RequestFailedException {
-    throw new RuntimeException("not implemented");
+    return requests.flatMap(this::validateRequest);
+  }
+
+  private Mono<ValidationResponse> validateRequest(ValidationRequest request) {
+    return switch (request.requestBody()) {
+      // Sample Registration
+      case SampleRegistrationInformation req ->
+          validateSampleMetadata(req, request.requestId(), request.projectId());
+      // Sample Update
+      case SampleUpdateInformation req ->
+          validateSampleMetadataUpdate(req, request.requestId(), request.projectId());
+      // Measurement Registration - NGS
+      case MeasurementRegistrationInformationNGS req ->
+          validateMeasurementMetadataNGS(req, request.requestId(), request.projectId());
+      // Measurement Update - NGS
+      case MeasurementUpdateInformationNGS req ->
+          validateMeasurementMetadataNGSUpdate(req, request.requestId(), request.projectId());
+      // Measurement Registration - Proteomics
+      case MeasurementRegistrationInformationPxP req ->
+          validateMeasurementMetadataPxP(req, request.requestId(), request.projectId());
+      // Measurement Update - Proteomics
+      case MeasurementUpdateInformationPxP req ->
+          validateMeasurementMetadataPxPUpdate(req, request.requestId(), request.projectId());
+      default -> Mono.error(new RequestFailedException("Invalid request"));
+    };
+  }
+
+
+  /**
+   * Ensures that the security context is applied in the correct order and written when it is
+   * required.
+   * <p>
+   * Also ensures that the {@link Callable} is executed on the {@link VirtualThreadScheduler} with
+   * {@link Mono#subscribeOn(Scheduler)}.
+   *
+   * @param securityApplicant
+   * @param serviceCallable
+   * @param converter
+   * @param transformer
+   * @return a {@link Mono} containing the
+   * {@link life.qbic.projectmanagement.application.api.AsyncProjectService.ValidationResponse}
+   */
+  private Mono<ValidationResponse> validateMetadata(
+      UnaryOperator<Mono<ValidationResponse>> securityApplicant,
+      Callable<ValidationResult> serviceCallable,
+      Function<ValidationResult, ValidationResponse> converter,
+      Function<Mono<ValidationResponse>, Publisher<ValidationResponse>> transformer) {
```

Comment:
The applicant, transformator and converter feel very unspecific. I would like to talk about that before approving the pull request.

Reply:
if it was specific, there would be no need to have this method, wouldn't it? Before the method extraction, there was a lot of code duplication for the currently four validations, with only very slightly differences based on the validation context (the actual detailed service calls).

Reply:
I get the method extraction. This is useful and I like it. However maybe we could do something like the following.
```java
    return usingSecurityContext(Mono.fromCallable(serviceCallable)
        .map(converter))
      .subscribeOn(scheduler);
```
The context has to be written in the calling code. No need to put it as a method argument. The caller has to ensure that the correct context is provided.
The context is not the responsibility of this method (as indicated by your method parameter)